### ISSUE TITLE
BUG: smithy spuriously dropping valid configurations

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/conda-smithy#2012]
This has been one of the most painful debugging sessions in a long time - the bug went away with random debug prints at random points in the code, even if those prints were long _after_ the (presumed) point of failure.

The situation is as follows, wanting to add another configuration to [clang-win-activation](https://github.com/conda-forge/clang-win-activation-feedstock/blob/main/recipe/conda_build_config.yaml), like so:
```diff
 CLANG_VERSION:
   - 16.0.6
   - 17.0.6
   - 18.1.8
+  - 19.1.0.rc1
```

the combination of conda-smithy 3.37.2 and conda-build 24.7.1 would spuriously drop the builds for 16.0.6, rather than doing something purely additively. Adding a 5th line would bring back all 5 variants, but 4 somehow means only 3 variants get created. There's also some zipping involved, so I put the exact commit I was rerendering (over and over and over) [here](https://github.com/h-vetinari/clang-win-activation-feedstock/commit/d5545be136222cb40a072c9bd2f1bfbf9ee9e32c).

I ultimately managed to get this to render correctly without debug prints (on top of 3.37.2 & `main`) with the following diff:
```diff
--- a/conda_smithy/configure_feedstock.py
+++ b/conda_smithy/configure_feedstock.py
@@ -983,7 +983,7 @@ def _render_ci_provider(

         config = conda_build.config.get_or_merge_config(
             None,
-            exclusive_config_file=forge_config["exclusive_config_file"],
+            exclusive_config_file=deepcopy(forge_config["exclusive_config_file"]),
             platform=platform,
             arch=arch,
         )
```
Since this happens at the interface between smithy and conda-build, I cannot tell who is responsible here; whether conda-build incorrectly mutates state that's passed in, or whether smithy incorrectly assumes that state will not be mutated.

It's also possible that my diagnosis is still not correct, and that the `deepcopy` just happen to randomly enforce some collision/synchronization that's not there otherwise, but doesn't (fully) get rid of the bug yet (like the print statements before; sidenote, there was often - but not always - a difference that `pprint.pprint` would remove the bug, but bare `print` wouldn't).

**Comment by h-vetinari** [comment]
@beckermr, any thoughts about this one? Would that patch make sense in your eyes? TBH I don't see how this could affect things (in the sense that `forge_config["exclusive_config_file"]` is just the filename, not the content), but at least it was 100% reproducible.

**Comment by beckermr** [comment]
Did you fix the pythonhashseed when debugging?

**Comment by h-vetinari** [comment]
Didn't need to - every run was perfectly reproducible in itself, even though there seemed to be no rhyme or reason which specific debug change caused it.

**Comment by beckermr** [comment]
Well as we all know computers are not magic. Deep copying a string should do nothing. 

Can you try a cast via "str(...)"?

Are we sure that dict entry is a string and not a Path object?

**Comment by beckermr** [comment]
The debug prints fixing things versus not also seems like a red herring.

**Comment by beckermr** [comment]
To directly answer you, that patch makes no sense to me at all.

**Comment by h-vetinari** [comment]
> Well as we all know computers are not magic.

To that I counter "Any sufficiently advanced technology is indistinguishable from magic." 😛 

IOW, just because it's reproducible doesn't meant it isn't (close to) magic. I wish python had a way to opt into pure functions... mutating self-referential state creates mind-bending puzzles. 😑 

> To directly answer you, that patch makes no sense to me at all.

Fair enough, I did point out in the OP that that's a distinct possibility (and that was even before I understood that the copied variable is just a string).

> Can you try a cast via "str(...)"?

With 
```diff
--- a/conda_smithy/configure_feedstock.py
+++ b/conda_smithy/configure_feedstock.py
@@ -983,7 +983,7 @@ def _render_ci_provider(

         config = conda_build.config.get_or_merge_config(
             None,
-            exclusive_config_file=forge_config["exclusive_config_file"],
+            exclusive_config_file=str(forge_config["exclusive_config_file"]),
             platform=platform,
             arch=arch,
         )
```
on top of 3.37.2, things fail again 🙃

**Comment by beckermr** [comment]
Yeah that's just crazy town right there. Whatever is happening it is subtle. 

I don't think we should patch just yet only because I don't think we've actually found the bug. 

Hrmmmmm.

**Comment by h-vetinari** [comment]
The plot thickens further... in https://github.com/conda-forge/ctng-compilers-feedstock/pull/148, even the bot gets it wrong (and the above patch doesn't help).

I parked a [branch](https://github.com/h-vetinari/ctng-compilers-feedstock/tree/bug) with that commit on my fork. The issue is that the rerender sets the `c_stdlib_version` to 2.17 universally in the variant configs, even though the CBC clearly [sets](https://github.com/h-vetinari/ctng-compilers-feedstock/blob/67452286ddef43215c7ac3a952623b03f815708f/recipe/conda_build_config.yaml#L34) one member of the key to 2.12.

**Comment by beckermr** [comment]
The value is not coming out of the function `_conda_build_api_render_for_smithy` in the list of metas so the bug is pretty low-level.

**Comment by h-vetinari** [comment]
Yeah, just got something similar. I put the following debug statements in (among others),
```diff
--- a/conda_smithy/configure_feedstock.py
+++ b/conda_smithy/configure_feedstock.py
@@ -1094,6 +1128,9 @@ def _render_ci_provider(
             if os.path.exists(_recipe_cbc + ".conda.smithy.bak"):
                 os.rename(_recipe_cbc + ".conda.smithy.bak", _recipe_cbc)

+        print("individual configs")
+        print(sorted(list({el["c_stdlib_version"] for meta in metas for el in meta[0].config.variants})))
+
         # render returns some download & reparsing info that we don't care about
         metas = [m for m, _, _ in metas]

@@ -1134,6 +1171,9 @@ def _render_ci_provider(
         fancy_platforms = []
         unfancy_platforms = set()

+        print("merged configs")
+        pprint.pprint(sorted(list({el["c_stdlib_version"] for metas in metas_list_of_lists for meta in metas for el in meta.config.variants})))
+
         configs = []
         for metas, platform, arch, enable, upload in zip(
             metas_list_of_lists,
```

and the result was
```
individual configs
['2.17']
individual configs
['2.17']
individual configs
['2.17']
individual configs
['2.17']
individual configs
[]                     <---- !
individual configs
['12']
merged configs
['12', '2.17']         <--- missing 2.12
```

**Comment by h-vetinari** [comment]
FWIW, I had tried the clang-win one with conda-build 24.5 out of curiosity (seeing that I just ran into a regression w.r.t. CBC-handling in 24.7 today), but the result was the same.

**Comment by beckermr** [comment]
The bug is here: https://github.com/conda-forge/conda-smithy/blob/main/conda_smithy/configure_feedstock.py#L895

For whatever reason, conda-build is marking the missing variant as skipped.

**Comment by beckermr** [comment]
The original recipe has a jinja2 if statement around the skip:

```
  {% if clang_major|int == 16 and cl_minor|int >= 40 %}
  # VS 17.10 requires clang>=17
  skip: true
  {% endif %}
```

Maybe we convert that to selectors and try again?

**Comment by beckermr** [comment]
the default jinja2 values are also suspicious:

```
{% set CLANG_VERSION = "16.0.6" %}
{% set CL_VERSION = "19.29" %}
```

**Comment by h-vetinari** [comment]
> Maybe we convert that to selectors and try again?

The problem is I didn't manage to make that combination work with selectors...

> the default jinja2 values are also suspicious:

why?

**Comment by beckermr** [comment]
So for selectors, this works fine, but is ofc hard coded:

```
skip: true  # [CLANG_VERSION == "16.0.6" and CL_VERSION == "19.40.33808"]
```

The ordering of when conda-build does the selectors vs jinja2 and it fills in values for jinja2 vars that are missing is a mystery to me. We shouldn't have to set defaults and yet we have to a lot. I am wondering if the ordering of when jinja2 is evaluated is making it so we need the default values and so is causing the skip to be added and then for w/e reason conda-build never tries again with non-default values.

**Comment by beckermr** [comment]
Adding a single random selector on some line like this

```
# [CLANG_VERSION and CL_VERSION]
```

causes things to work again.

**Comment by beckermr** [comment]
Also smithy fails to render without the defaults, ending in an error inside conda-build. So it is not only the linter that needs the defaults.

**Comment by h-vetinari** [comment]
> The original recipe has a jinja2 if statement around the skip:

FWIW, the ctng-compilers recipe has no such jinja-skip. Perhaps the issues are separate though.

**Comment by beckermr** [comment]
I think the easiest path forward is to change the CBC to zip the version parts you need into the variant variables and use them in the selectors.

**Comment by beckermr** [comment]
One common thing is that both recipes set jinja2 default values that are also set in the variant configs. I wonder if that is causing wires to cross.

**Comment by h-vetinari** [comment]
As a temporary workaround or as a longer-term fix? I'd be fine with the former, not excited about the latter (not least because this used to work, and it's a legitimate use of the various mechanisms).

**Comment by beckermr** [comment]
Can you find a combination of current smithy and old conda-build where this does work? That would help eliminate sources of bugs.

**Comment by beckermr** [comment]
@h-vetinari here is a minimal reproducer 

https://github.com/conda/conda-build/issues/5445

This confirms the bug is in conda-build, not smithy.

**Comment by h-vetinari** [comment]
Thank you so much for chasing down that bug and coming up with a fix already! 🙏😊

**Comment by h-vetinari** [comment]
One thing it does confirm though is that the issue in ctng-compilers is a separate one; I tried a local install of https://github.com/conda/conda-build/pull/5447, and it still incorrectly bumps the `c_stdlib_version`. 😑

**Comment by beckermr** [comment]
Oh boy yay fun! We'll have to keep digging.

**Comment by h-vetinari** [comment]
I tried to reduce to an example like you did in https://github.com/conda/conda-build/issues/5445, but I still see 2.12 in the `conda rerender .` output, even though it doesn't show in the variant configs if I run smithy. 🤔 

So it looks like this is at least _partly_ smithy's fault... Indeed, I had been looking at the wrong variable [above](https://github.com/conda-forge/conda-smithy/issues/2012#issuecomment-2270447289); if I do the same with `cross_target_stdlib_version`, I get:

```
individual configs
['12', '2.12', '2.17']
individual configs
['12', '2.12', '2.17']
individual configs
['12', '2.12', '2.17']
individual configs
['12', '2.12', '2.17']
individual configs
[]                        # presumably this is osx, which we're skipping
individual configs
['12']
merged configs
['12', '2.12', '2.17']
```

Chasing this down a bit further, we seem to be dropping `cross_target_stdlib_version` somewhere in this block:
https://github.com/conda-forge/conda-smithy/blob/66d72937398aa9d2fd651fda90cb764440662e9d/conda_smithy/configure_feedstock.py#L203-L253

IOW, after that block `squished_variants` does not contain that key anymore (except as part of the zip specification)

```
OrderedDict([('openmp_ver', ['4.5']),
             ('docker_image', ['quay.io/condaforge/linux-anvil-cos7-x86_64']),
             ('binutils_version', ['2.40']),
             ('channel_sources',
              ['conda-forge/label/sysroot-with-crypt,conda-forge']),
             ('libgomp_ver', ['1.0.0']),
             ('c_stdlib', ['sysroot']),
             ('channel_targets', ['conda-forge main']),
             ('c_stdlib_version', ('2.17',)),
             ('cdt_name', ('cos7',)),
             ('zip_keys',
              [['c_stdlib_version', 'cdt_name'],
               ['gcc_version', 'gcc_maj_ver', 'libgfortran_soname'],
               ['cross_target_stdlib_version',
                'cross_target_stdlib',
                'triplet',
                'old_triplet',
                'cross_target_platform']])])
```
whereas before that block it looks like:
```
OrderedDict([('openmp_ver', ['4.5']),
             ('target_platform', ['linux-s390x']),
             ('docker_image', ['quay.io/condaforge/linux-anvil-cos7-x86_64']),
             ('binutils_version', ['2.40']),
             ('channel_sources',
              ['conda-forge/label/sysroot-with-crypt,conda-forge']),
             ('libgomp_ver', ['1.0.0']),
             ('c_stdlib', ['sysroot']),
             ('channel_targets', ['conda-forge main']),
             ('gcc_version', ('12.4.0', '13.3.0', '14.1.0')),
             ('gcc_maj_ver', ('12', '13', '14')),
             ('libgfortran_soname', ('5', '5', '5')),
             ('c_stdlib_version', ('2.17',)),
             ('cdt_name', ('cos7',)),
             ('cross_target_stdlib_version',
              ('2.17', '2.12', '2.17', '12', '2.17')),                       <--- SEE HERE
             ('old_triplet',
              ('s390x-conda-linux-gnu',
               'x86_64-conda_cos6-linux-gnu',
               'aarch64-conda_cos7-linux-gnu',
               'x86_64-w64-mingw32',
               'powerpc64le-conda_cos7-linux-gnu')),
             ('cross_target_platform',
              ('linux-s390x',
               'linux-64',
               'linux-aarch64',
               'win-64',
               'linux-ppc64le')),
             ('triplet',
              ('s390x-conda-linux-gnu',
               'x86_64-conda-linux-gnu',
               'aarch64-conda-linux-gnu',
               'x86_64-w64-mingw32',
               'powerpc64le-conda-linux-gnu')),
             ('cross_target_stdlib',
              ('sysroot', 'sysroot', 'sysroot', 'm2w64-sysroot', 'sysroot')),
             ('zip_keys',
              [['c_stdlib_version', 'cdt_name'],
               ['gcc_version', 'gcc_maj_ver', 'libgfortran_soname'],
               ['cross_target_stdlib_version',
                'cross_target_stdlib',
                'triplet',
                'old_triplet',
                'cross_target_platform']])])
```

**Comment by h-vetinari** [comment]
Nevermind, that was a misdiagnosis, the bug happens later still. After the block I mentioned, the correct config is still found in `zipped_configs`:
```
[[OrderedDict([('cross_target_stdlib_version', ['12']),
               ('cross_target_stdlib', ['m2w64-sysroot']),
               ('triplet', ['x86_64-w64-mingw32']),
               ('old_triplet', ['x86_64-w64-mingw32']),
               ('cross_target_platform', ['win-64'])]),
  OrderedDict([('cross_target_stdlib_version', ['2.17']),
               ('cross_target_stdlib', ['sysroot']),
               ('triplet', ['powerpc64le-conda-linux-gnu']),
               ('old_triplet', ['powerpc64le-conda_cos7-linux-gnu']),
               ('cross_target_platform', ['linux-ppc64le'])]),
  OrderedDict([('cross_target_stdlib_version', ['2.12']),                          <-----
               ('cross_target_stdlib', ['sysroot']),
               ('triplet', ['x86_64-conda-linux-gnu']),
               ('old_triplet', ['x86_64-conda_cos6-linux-gnu']),
               ('cross_target_platform', ['linux-64'])]),
  OrderedDict([('cross_target_stdlib_version', ['2.17']),
               ('cross_target_stdlib', ['sysroot']),
               ('triplet', ['s390x-conda-linux-gnu']),
               ('old_triplet', ['s390x-conda-linux-gnu']),
               ('cross_target_platform', ['linux-s390x'])]),
  OrderedDict([('cross_target_stdlib_version', ['2.17']),
               ('cross_target_stdlib', ['sysroot']),
               ('triplet', ['aarch64-conda-linux-gnu']),
               ('old_triplet', ['aarch64-conda_cos7-linux-gnu']),
               ('cross_target_platform', ['linux-aarch64'])])],
 [OrderedDict([('gcc_version', ['14.1.0']),
               ('gcc_maj_ver', ['14']),
               ('libgfortran_soname', ['5'])]),
  OrderedDict([('gcc_version', ['12.4.0']),
               ('gcc_maj_ver', ['12']),
               ('libgfortran_soname', ['5'])]),
  OrderedDict([('gcc_version', ['13.3.0']),
               ('gcc_maj_ver', ['13']),
               ('libgfortran_soname', ['5'])])]]
```

**Comment by h-vetinari** [comment]
UGH, the solution is: the rerender works, and I managed to confuse myself in a cross-compilation recipe. What was happening is that we were doing the first rerender since we dropped cos6, and so the `c_stdlib_version` coming in from the global pinning got bumped (but _not_ the `cross_target_stdlib_version` that's specified in the CBC).

Sorry about the hassle here. I think we can close this since we've identified the error for clang-win-activation, realised my error for ctng-compilers. That leaves just https://github.com/conda/conda-build/issues/5443 of the recent compiler-related rendering regressions unresolved 😅
